index.js

```diff
@@ -1020,46 +1030,78 @@ module.exports = class Autobase extends ReadyResource {
     await this._makeLinearizer(this.system)
   }
 
-  // NOTE: runs in parallel with everything, can never fail
+  async initialFastForward (key, length, timeout) {
+    const core = this.store.get(key)
+    const target = await this._preFastForward(core, length, timeout)
+
+    // initial fast-forward failed
+    if (target === null) {
+      this.prologue = null
+      return
+    }
+
+    for (const w of this.activeWriters) w.pause()
+
+    this.fastForwarding = false
+    this.fastForwardTo = target
+
+    this._bumpAckTimer()
+    this._queueBump()
+  }
+
   async queueFastForward () {
     // if already FFing, let the finish. TODO: auto kill the attempt after a while and move to latest?
-    if (this.fastForwardingTo !== 0) return
+    if (!this.fastForwardEnabled || this.fastForwarding) return
 
     const core = this.system.core.getBackingCore()
 
     if (core.session.length <= core.length + FF_THRESHOLD) return
     if (this.fastForwardTo !== null && core.session.length <= this.fastForwardTo.length + FF_THRESHOLD) return
 
-    this.fastForwardingTo = core.session.length
+    const target = await this._preFastForward(core.session, core.session.length, null)
+
+    // fast-forward failed
+    if (target === null) return
+
+    // if it migrated underneath us, ignore for now
+    if (core !== this.system.core.getBackingCore()) return
+
+    for (const w of this.activeWriters) w.pause()
+
+    this.fastForwardTo = target
+
     this._bumpAckTimer()
+    this._queueBump()
+  }
 
-    const fastForwardTo = {
-      key: core.session.key,
-      length: this.fastForwardingTo,
-      migrate: null
-    }
+  // NOTE: runs in parallel with everything, can never fail
+  async _preFastForward (core, length, timeout) {
+    this.fastForwarding = true
+
+    const info = { key: core.key, length }
 
     try {
       // sys runs open with wait false, so get head block first for low complexity
-      if (!(await core.session.has(this.fastForwardingTo - 1))) {
-        await core.session.get(this.fastForwardingTo - 1)
+      if (!(await core.has(length - 1))) {
+        await core.get(length - 1, { timeout })
       }
 
-      const system = new SystemView(core.session.session(), this.fastForwardingTo)
+      const system = new SystemView(core.session(), length)
       await system.ready()
 
       if (system.version > this.maxSupportedVersion) {
         const upgrade = {
           version: system.version,
-          length: this.fastForwardingTo
+          length
         }
 
-        this.fastForwardingTo = 0
+        this.fastForwarding = false
         this.emit('upgrade-available', upgrade)
-        return
+        return null
       }
 
-      const migrated = !system.sameIndexers(this.linearizer.indexers)
+      const systemShouldMigrate = b4a.equals(core.key, this.system.core.key) &&
+        !system.sameIndexers(this.linearizer.indexers)
```

Comment:
technically we need to check the prologue also here? ie add X, remove X, add X

Reply:
that would only be the case for double ff, but in that case the first part of the condition would be false.

this is only there to check if we are at the very